### PR TITLE
bpf: add raw probes for verifier capabilities

### PR DESCRIPTION
**Pull request by borkmann**
This work adds a new probe framework for testing the verifier from
raw eBPF byte code. This is needed because some tests cannot be done
through clang, but require specific byte code generation in order to
test the verifier for certain patterns.

We ship two samples, raw_change_tail.t and raw_mark_map_val.t that
also act as templates for future test cases. Generated header when
both tests pass:

  $ ./run_probes.sh . /tmp/
  $ cat /tmp/globals/bpf_features.h
  #ifndef BPF_FEATURES_H_
  #define BPF_FEATURES_H_

  #define HAVE_SKB_CHANGE_TAIL

  #define HAVE_MARK_MAP_VALS

  #endif /* BPF_FEATURES_H_ */

Generated header when not all tests pass, where the verifier output
is dumped in #if 0 / #endif section for debugging:

  $ ./run_probes.sh . /tmp/
  $ cat /tmp/globals/bpf_features.h
  #ifndef BPF_FEATURES_H_
  #define BPF_FEATURES_H_

  // #define HAVE_SKB_CHANGE_TAIL

  #if 0
  HAVE_SKB_CHANGE_TAIL failed due to load error: Invalid argument
  0: (b7) r2 = 0
  1: (b7) r3 = 0
  2: (85) call 38
  invalid func 38
  #endif

  // #define HAVE_MARK_MAP_VALS

  #if 0
  HAVE_MARK_MAP_VALS failed due to load error: Permission denied
  0: (bf) r2 = r10
  1: (07) r2 += -8
  2: (7a) *(u64 *)(r2 +0) = 0
  3: (18) r1 = 0xd8070cc0
  5: (85) call 1
  6: (bf) r1 = r10
  7: (07) r1 += -152
  8: (7b) *(u64 *)(r1 +0) = r0
  9: (15) if r0 == 0x0 goto pc+2
   R0=map_value(ks=8,vs=8) R1=fp-152 R10=fp fp-152=map_value_or_null
  10: (79) r3 = *(u64 *)(r1 +0)
  11: (7a) *(u64 *)(r3 +0) = 42
  R3 invalid mem access 'map_value_or_null'
  #endif
  #endif /* BPF_FEATURES_H_ */

Newly added test cases just need to add the *.t file to probes, where
run_probes.sh will pick them up automatically. This also removes the
dependency on tc's ability to parse BPF ELF files for test cases. The
HAVE_MARK_MAP_VALS will be instrumented in another commit.

Related: #310

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/cilium/cilium/pull/380%23issuecomment-288075276%22%2C%20%22https%3A//github.com/cilium/cilium/pull/380%23issuecomment-288088728%22%2C%20%22https%3A//github.com/cilium/cilium/pull/380%23issuecomment-288090257%22%2C%20%22https%3A//github.com/cilium/cilium/pull/380%23discussion_r107177214%22%2C%20%22https%3A//github.com/cilium/cilium/pull/380%23discussion_r107178811%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/380%23issuecomment-288075276%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Fyi%2C%20I%20did%20extend%20the%20original%20patch%20in%20order%20to%20add%20a%20log%20notification%20to%20the%20user.%22%2C%20%22created_at%22%3A%20%222017-03-21T13%3A19%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40tgraf%20Yes%2C%20that%20is%20correct.%20I%20took%20the%20liberty%20to%20remove%20the%20skb_change_tail.c%20test%2C%20but%20I%20left%20the%20shell%20function%20still%20in%20tact%20in%20case%20we%20might%20need%20it%20in%20future.%20But%20if%20preferred%20I%20can%20also%20remove%20that%20part%20since%20we%20have%20it%20in%20the%20git%20history%20anyway.%22%2C%20%22created_at%22%3A%20%222017-03-21T14%3A07%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3E%20%40tgraf%20Yes%2C%20that%20is%20correct.%20I%20took%20the%20liberty%20to%20remove%20the%20skb_change_tail.c%20test%2C%20but%20I%20left%20the%20shell%20function%20still%20in%20tact%20in%20case%20we%20might%20need%20it%20in%20future.%20But%20if%20preferred%20I%20can%20also%20remove%20that%20part%20since%20we%20have%20it%20in%20the%20git%20history%20anyway.%5Cr%5Cn%5Cr%5CnThat%27s%20fine%2C%20I%20just%20wanted%20to%20confirm.%20Code%20works%20fine%20for%20me%20as-is.%22%2C%20%22created_at%22%3A%20%222017-03-21T14%3A12%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%209cb50e0401ac44705f936a700dfb90f59a108e2b%20daemon/daemon.go%2016%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/380%23discussion_r107177214%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20see%20this%20here%20but%20when%20is%20this%20%28%60bpf_features.log%60%29%20file%20created%3F%20And%20by%20whom%3F%22%2C%20%22created_at%22%3A%20%222017-03-21T14%3A55%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/5714066%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aanm%22%7D%7D%2C%20%7B%22body%22%3A%20%22%60%60%60%5Cr%5Cn%5Ct%5Ct%24OUT/%24OUT_BIN%201%3E%3E%20%24FEATURE_FILE%202%3E%3E%20%24WARNING_FILE%5Cr%5Cn%60%60%60%22%2C%20%22created_at%22%3A%20%222017-03-21T15%3A00%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20daemon/daemon.go%3AL238-253%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/cilium/cilium/pull/380#issuecomment-288075276'>General Comment</a></b>
- <a href='https://github.com/borkmann'></a> Fyi, I did extend the original patch in order to add a log notification to the user.
- <a href='https://github.com/borkmann'></a> @tgraf Yes, that is correct. I took the liberty to remove the skb_change_tail.c test, but I left the shell function still in tact in case we might need it in future. But if preferred I can also remove that part since we have it in the git history anyway.
- <a href='https://github.com/tgraf'></a> > @tgraf Yes, that is correct. I took the liberty to remove the skb_change_tail.c test, but I left the shell function still in tact in case we might need it in future. But if preferred I can also remove that part since we have it in the git history anyway.
That's fine, I just wanted to confirm. Code works fine for me as-is.
- [ ] <a href='#crh-comment-Pull 9cb50e0401ac44705f936a700dfb90f59a108e2b daemon/daemon.go 16'></a> &nbsp;<b><a href='https://github.com/cilium/cilium/pull/380#discussion_r107177214'>File: daemon/daemon.go:L238-253</a></b>
- <a href='https://github.com/aanm'></a> I see this here but when is this (`bpf_features.log`) file created? And by whom?
- <a href='https://github.com/tgraf'></a> 
```
$OUT/$OUT_BIN 1>> $FEATURE_FILE 2>> $WARNING_FILE
```

<a href='https://www.codereviewhub.com/cilium/cilium/pull/380?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/cilium/cilium/pull/380?approve=1'></a>&nbsp;<a href='https://github.com/cilium/cilium/pull/380'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by aanm** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/cilium/cilium/pull/380'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by borkmann** [comment]
Fyi, I did extend the original patch in order to add a log notification to the user.

**Review by tgraf — APPROVED** [review]
Looks awesome. My read is that we don't nee the existing probe to detect skb_change_tail() at all anymore, correct?

**Comment by borkmann** [comment]
@tgraf Yes, that is correct. I took the liberty to remove the skb_change_tail.c test, but I left the shell function still in tact in case we might need it in future. But if preferred I can also remove that part since we have it in the git history anyway.

**Comment by tgraf** [comment]
> @tgraf Yes, that is correct. I took the liberty to remove the skb_change_tail.c test, but I left the shell function still in tact in case we might need it in future. But if preferred I can also remove that part since we have it in the git history anyway.

That's fine, I just wanted to confirm. Code works fine for me as-is.

File: daemon/daemon.go [hunk 1]
```diff
@@ -237,9 +238,15 @@ func (d *Daemon) compileBase() error {
 		log.Warningf("Command output:\n%s", out)
 		return err
 	}
+
 	log.Warningf("Enabled bpf_jit_enable")
 	log.Warningf("Disabled rp_filter on all interfaces!")
 
+	fout, err := ioutil.ReadFile(filepath.Join(d.conf.RunDir, "bpf_features.log"))
```
Comment (aanm) [inline daemon/daemon.go:245]: I see this here but when is this (`bpf_features.log`) file created? And by whom?

Comment (tgraf) [reply daemon/daemon.go:245]:
```
		$OUT/$OUT_BIN 1>> $FEATURE_FILE 2>> $WARNING_FILE
```

**Pull request merged**